newrelic/hooks/mlmodel_gemini.py

```diff
@@ -0,0 +1,204 @@
+    transaction.add_ml_model_info("Gemini", GEMINI_VERSION)
+    transaction._add_agent_attribute("llm", True)
+
+    # Obtain attributes to be stored on embedding events regardless of whether we hit an error
+    embedding_id = str(uuid.uuid4())
+
+    ft = FunctionTrace(name=wrapped.__name__, group="Llm/embedding/Gemini")
+    ft.__enter__()
+    linking_metadata = get_trace_linking_metadata()
+    try:
+        response = wrapped(*args, **kwargs)
+    except Exception as exc:
+        _record_embedding_error(transaction, embedding_id, linking_metadata, kwargs, ft, exc)
+        raise
+    ft.__exit__(None, None, None)
+
+    if not response:
+        return response
+
+    _record_embedding_success(transaction, embedding_id, linking_metadata, kwargs, ft)
+    return response
+
+
+async def wrap_embed_content_async(wrapped, instance, args, kwargs):
+    transaction = current_transaction()
+    if not transaction:
+        return await wrapped(*args, **kwargs)
+
+    settings = transaction.settings if transaction.settings is not None else global_settings()
+    if not settings.ai_monitoring.enabled:
+        return await wrapped(*args, **kwargs)
+
+    # Framework metric also used for entity tagging in the UI
+    transaction.add_ml_model_info("Gemini", GEMINI_VERSION)
+    transaction._add_agent_attribute("llm", True)
+
+    # Obtain attributes to be stored on embedding events regardless of whether we hit an error
+    embedding_id = str(uuid.uuid4())
+
+    ft = FunctionTrace(name=wrapped.__name__, group="Llm/embedding/Gemini")
+    ft.__enter__()
+    linking_metadata = get_trace_linking_metadata()
+    try:
+        response = await wrapped(*args, **kwargs)
+    except Exception as exc:
+        _record_embedding_error(transaction, embedding_id, linking_metadata, kwargs, ft, exc)
+        raise
+    ft.__exit__(None, None, None)
+
+    if not response:
+        return response
+
+    _record_embedding_success(transaction, embedding_id, linking_metadata, kwargs, ft)
+    return response
+
+
+def _record_embedding_error(transaction, embedding_id, linking_metadata, kwargs, ft, exc):
+    settings = transaction.settings if transaction.settings is not None else global_settings()
+    span_id = linking_metadata.get("span.id")
+    trace_id = linking_metadata.get("trace.id")
+
+    notice_error_attributes = {}
+    try:
+        embedding_content = kwargs.get("contents")
+        embedding_content = str(embedding_content) if embedding_content else ""
+        model = kwargs.get("model")
+
+        notice_error_attributes = {
+            "http.statusCode": getattr(exc, "code", None),
+            "error.message": getattr(exc, "message", None),
+            "error.code": getattr(exc, "status", None),  # ex: 'NOT_FOUND'
+            "embedding_id": embedding_id,
+        }
+    except Exception:
+        _logger.warning(EXCEPTION_HANDLING_FAILURE_LOG_MESSAGE, traceback.format_exception(*sys.exc_info()))
```

Comment:
Is there a difference between this and `_logger.warning(EXCEPTION_HANDLING_FAILURE_LOG_MESSAGE, exc_info=True)`?

Reply:
Tested this and they are functionally equivalent so I'll swap to use `exc_info=True`